### PR TITLE
Implement efficient way to fetch partition name and values

### PR DESCRIPTION
**Pull request by jzzhaofb**

### CODE REVIEW

**Comment by jainxrohit** [comment]
Can you please write few tests in the TestHiveUtils to verify it?

**Comment by jainxrohit** [comment]
Commit messages should follow these guidelines (http://chris.beams.io/posts/git-commit/)

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 1]
```diff
@@ -469,12 +469,38 @@ public static void renameFile(FileSystem fileSystem, Path source, Path target)
     // TODO: https://github.com/prestodb/presto/issues/15974
     public static Map<String, String> toPartitionNamesAndValues(String partitionName)
     {
-        try {
-            return ImmutableMap.copyOf(makeSpecFromName(partitionName));
+        ImmutableMap.Builder<String,String> resultBuilder = ImmutableMap.builder();
+        int index = 0;
+        int len = partitionName.length();
```
Comment (jainxrohit) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:474]: Presto codebase guidelines recommends using complete names. Hence lets call it length.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 2]
```diff
@@ -469,12 +469,38 @@ public static void renameFile(FileSystem fileSystem, Path source, Path target)
...
+            }
+            
+            int keyEnd = index++;
+            int valueStart = index;
+            while(index<len && partitionName.charAt(index) != '/') {
+                index++;
+            }
+            int valueEnd = index++;
+
+            String key = partitionName.substring(keyStart,keyEnd);
```
Comment (jainxrohit) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:494]: use unescapePathName

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 3]
```diff
@@ -469,12 +469,38 @@ public static void renameFile(FileSystem fileSystem, Path source, Path target)
...
+        int index = 0;
+        int len = partitionName.length();
+        boolean invalidName = false;
+
+        while(index < len){
+            int keyStart = index;
+            while(index < len && partitionName.charAt(index) != '=') {
+                index++;
+            }
+            if(index >= len){
```
Comment (jainxrohit) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:482]: Replace it with checkState(index < len, "Invalid partition spec: " + partitionName);

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 4]
```diff
@@ -469,12 +469,38 @@ public static void renameFile(FileSystem fileSystem, Path source, Path target)
     // TODO: https://github.com/prestodb/presto/issues/15974
     public static Map<String, String> toPartitionNamesAndValues(String partitionName)
     {
-        try {
-            return ImmutableMap.copyOf(makeSpecFromName(partitionName));
+        ImmutableMap.Builder<String,String> resultBuilder = ImmutableMap.builder();
+        int index = 0;
+        int len = partitionName.length();
+        boolean invalidName = false;
```
Comment (jainxrohit) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:475]: Not required.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 5]
```diff
@@ -469,12 +469,38 @@ public static void renameFile(FileSystem fileSystem, Path source, Path target)
...
+            }
+            int valueEnd = index++;
+
+            String key = partitionName.substring(keyStart,keyEnd);
+            String value = partitionName.substring(valueStart,valueEnd);
+            resultBuilder.put(key,value);
         }
-        catch (MetaException e) {
+
+        if(invalidName){
```
Comment (jainxrohit) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:499]: Not required after using checkState

**Review by jainxrohit — CHANGES_REQUESTED** [review]
Please also squash all commits in to one commit.

**Comment by jainxrohit** [comment]
Use the commit message and PR title as
Implement efficient way to fetch partition name and values

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveMaterializedViewUtils.java [hunk 6]
```diff
@@ -95,6 +95,79 @@ public void testMaterializedDataPredicates()
         comparePredicates(materializedDataPredicates, keys, partitionResults.build());
     }
 
+    @Test
+    public void testMaterializedDataPredicatesWithSingleColumn()
```
[the remark this answers is not in the document]
Comment (jainxrohit) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveMaterializedViewUtils.java:99]: By moving to TestHiveUtil, you can simplify many of these details.

File: presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java [hunk 7]
```diff
@@ -471,35 +471,27 @@ public static void renameFile(FileSystem fileSystem, Path source, Path target)
     {
```
Comment (jainxrohit) [inline presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java:471]: Remove the TODO

**Review by jainxrohit — CHANGES_REQUESTED** [review]

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java [hunk 8]
```diff
@@ -42,6 +44,8 @@
 import static com.facebook.presto.hive.HiveUtil.parseHiveTimestamp;
 import static com.facebook.presto.hive.HiveUtil.shouldUseRecordReaderFromInputFormat;
 import static com.facebook.presto.hive.metastore.MetastoreUtil.toPartitionValues;
+import static com.facebook.presto.hive.metastore.MetastoreUtil.toPartitionNamesAndValues;
+import static org.apache.hadoop.hive.common.FileUtils.unescapePathName;
```
Comment (jainxrohit) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java:48]: remove unused imports. Make sure to build(mvn install) locally to verify.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java [hunk 9]
```diff
@@ -116,6 +144,38 @@ private static void assertToPartitionValues(String partitionName)
...
+                String actualValue = actual.get(key);
+                String expectedValue = expectedValueList.get(index);
+                if (!actualValue.equals(expectedValue)) {
+                    equalMaps = false;
+                    break;
+                }
+            }
+        }
+
+        assertTrue(equalMaps);
```
Comment (jainxrohit) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java:176]: rather checking in the end, may we should assert it when the equal fails at first spot?

**Review by jainxrohit — CHANGES_REQUESTED** [review]

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java [hunk 10]
```diff
@@ -116,6 +142,39 @@ private static void assertToPartitionValues(String partitionName)
...
+                }
+                String actualValue = actual.get(key);
+                String expectedValue = expectedValueList.get(index);
+                if (!actualValue.equals(expectedValue)) {
+                    assertTrue(false);
+                    break;
+                }
+            }
+
+            assertTrue(true);
```
Comment (jainxrohit) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java:172]: This is redundant.

**Review by jainxrohit — CHANGES_REQUESTED** [review]

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java [hunk 11]
```diff
@@ -116,6 +142,39 @@ private static void assertToPartitionValues(String partitionName)
...
+    {
+        Map<String,String> actual = toPartitionNamesAndValues(partitionName);
+        AbstractList<String> expectedValueList = new ArrayList<>();
+        for( String s: expectedKeyList) {
+            expectedValueList.add(null);
+        }
+        Warehouse.makeValsFromName(partitionName, expectedValueList);
+
+        if(actual.keySet().size() != expectedKeyList.size()){
+            assertTrue(false);
```
Comment (jainxrohit) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveUtil.java:156]: replace this with 
`checkstate(actual.keySet().size() == expectedKeyList.size())`

Similarly at other place.

**Review by jainxrohit — CHANGES_REQUESTED** [review]

**Review by jainxrohit — APPROVED** [review]
LGTM

**Review by highker — APPROVED** [review]

**Pull request merged**